druid-shell/src/gtk/mod.rs

```diff
@@ -0,0 +1,689 @@
+};
+use piet_common::{Piet, RenderContext};
+use util::assert_main_thread;
+use win_main::with_application;
+
+use crate::keyboard;
+use crate::kurbo::{Point, Vec2};
+use crate::platform::dialog::{FileDialogOptions, FileDialogType};
+use crate::platform::menu::Menu;
+use crate::window::{self, Cursor, MouseButton, Text, TimerToken, WinCtx, WinHandler};
+use crate::Error;
+
+pub mod application;
+pub mod dialog;
+pub mod menu;
+pub mod util;
+pub mod win_main;
+
+#[derive(Clone, Default)]
+pub struct WindowHandle {
+    state: Weak<WindowState>,
+}
+
+/// Builder abstraction for creating new windows
+pub struct WindowBuilder {
+    handler: Option<Box<WinHandler>>,
+    title: String,
+    menu: Option<menu::Menu>,
+}
+
+#[derive(Clone)]
+pub struct IdleHandle {
+    idle_queue: Arc<Mutex<Vec<Box<dyn IdleCallback>>>>,
+    state: Weak<WindowState>,
+}
+
+// TODO: move this out of platform-dependent section.
+trait IdleCallback: Send {
+    fn call(self: Box<Self>, a: &Any);
+}
+
+impl<F: FnOnce(&Any) + Send> IdleCallback for F {
+    fn call(self: Box<F>, a: &Any) {
+        (*self)(a)
+    }
+}
+
+struct WindowState {
+    window: ApplicationWindow,
+    handler: RefCell<Box<dyn WinHandler>>,
+    idle_queue: Arc<Mutex<Vec<Box<dyn IdleCallback>>>>,
+    current_keyval: RefCell<Option<u32>>,
+}
+
+struct WinCtxImpl<'a> {
+    handle: &'a WindowHandle,
+    text: Text<'static>,
+}
+
+impl WindowBuilder {
+    pub fn new() -> WindowBuilder {
+        WindowBuilder {
+            handler: None,
+            title: String::new(),
+            menu: None,
+        }
+    }
+
+    pub fn set_handler(&mut self, handler: Box<WinHandler>) {
+        self.handler = Some(handler);
+    }
+
+    pub fn set_title(&mut self, title: impl Into<String>) {
+        self.title = title.into();
+    }
+
+    pub fn set_menu(&mut self, menu: Menu) {
+        self.menu = Some(menu);
+    }
+
+    pub fn build(self) -> Result<WindowHandle, Error> {
+        assert_main_thread();
+
+        let handler = self
+            .handler
+            .expect("Tried to build a window without setting the handler");
+
+        let window = with_application(|app| ApplicationWindow::new(&app));
+
+        let accel_group = AccelGroup::new();
+        window.add_accel_group(&accel_group);
+
+        let vbox = gtk::Box::new(gtk::Orientation::Vertical, 0);
+        window.add(&vbox);
+
+        let win_state = Arc::new(WindowState {
+            window,
+            handler: RefCell::new(handler),
+            idle_queue: Arc::new(Mutex::new(vec![])),
+            current_keyval: RefCell::new(None),
+        });
+
+        let state = win_state.clone();
+        win_state.window.connect_destroy(move |_| {
+            // this ties a clone of Arc<WindowState> to the ApplicationWindow to keep it alive
+            // when the ApplicationWindow is destroyed, the last Arc is dropped
+            // and any Weak<WindowState> will be None on upgrade()
+            let _ = &state;
```

Comment:
Are you sure this works? it feels a bit funny to me, would be curious if this idea came from somewhere?